src/Agent.Plugins/PipelineCache/TarUtils.cs

```diff
@@ -0,0 +1,232 @@
+
+namespace Agent.Plugins.PipelineCache
+{
+    public static class TarUtils
+    {
+        private readonly static bool isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+
+        private const string archiveFileName = "archive.tar";
+
+        /// <summary>
+        /// Will archive files in the input path into a TAR file.
+        /// </summary>
+        /// <returns>The path to the TAR.</returns>
+        public static async Task<string> ArchiveFilesToTarAsync(
+            AgentTaskPluginExecutionContext context,
+            string inputPath,
+            CancellationToken cancellationToken)
+        {
+            var archiveFile = Path.Combine(Path.GetTempPath(), archiveFileName);
+            if (File.Exists(archiveFile))
+            {
+                File.Delete(archiveFile);
+            }
+            var processFileName = "tar";
+            var processArguments = $"-cf {archiveFile} -C {inputPath} .";
+
+            Action actionOnFailure = () =>
+            {
+                // Delete archive file.
+                if (File.Exists(archiveFile))
+                {
+                    File.Delete(archiveFile);
+                }
+            };
+
+            await RunProcessAsync(
+                context,
+                processFileName,
+                processArguments,
+                // no additional tasks on create are required to run whilst running the TAR process
+                (Process process, CancellationToken ct) => Task.CompletedTask,
+                actionOnFailure,
+                cancellationToken);
+            return archiveFile;
+        }
+
+        /// <summary>
+        /// This will download the dedup into stdin stream while extracting the TAR simulataneously (piped). This is done by
+        /// starting the download through a Task and starting the TAR/7z process which is reading from STDIN.
+        /// </summary>
+        /// <remarks>
+        /// Windows will use 7z to extract the TAR file (only if 7z is installed on the machine and is part of PATH variables). 
+        /// Non-Windows machines will extract TAR file using the 'tar' command'.
+        /// </remarks>
+        public static Task DownloadAndExtractTarAsync(
+            AgentTaskPluginExecutionContext context,
+            Manifest manifest,
+            DedupManifestArtifactClient dedupManifestClient,
+            string targetDirectory,
+            CancellationToken cancellationToken)
+        {
+            ValidateTarManifest(manifest);
+            DedupIdentifier dedupId = DedupIdentifier.Create(manifest.Items.Single(i => i.Path == $"/{archiveFileName}").Blob.Id);
+            bool does7zExists = isWindows ? CheckIf7ZExists() : false;
+            string processFileName = (isWindows && does7zExists) ? "7z" : "tar";
+            string processArguments = (isWindows && does7zExists) ? $"x -si -aoa -o{targetDirectory} -ttar" : $"-xf - -C {targetDirectory}";
+
```

Comment:
string processFileName, processArguments;
if (does7zExist) {

} else {

}

Reply:
probably need to wrap {} targetDirectory in double quotes

---
In reply to: [315360013](https://github.com/microsoft/azure-pipelines-agent/pull/2369#discussion_r315360013) [](ancestors = 315360013)